feat: implement Redis Pub/Sub for real-time location update. Enhance type safety with shared libs/ws/index types.

jackrdye opened the pull request:
- Added Redis Pub/Sub functionality to the LocationService for publishing real-time location updates.
- Enhanced SocketHandler to subscribe to Redis channel and broadcast location updates to connected WebSocket clients.
- Implemented server heartbeat mechanism to maintain active WebSocket connections and clean up stale connections.
- Updated WebSocket message handling to include ping/pong for connection health checks.
- Create type safe websockets with shared types in libs/ws/index

jackrdye committed:
feat: implement Redis Pub/Sub for real-time location update. Enhance type safety with shared libs/ws/index types.

- Added Redis Pub/Sub functionality to the LocationService for publishing real-time location updates.
- Enhanced SocketHandler to subscribe to Redis channel and broadcast location updates to connected WebSocket clients.
- Implemented server heartbeat mechanism to maintain active WebSocket connections and clean up stale connections.
- Updated WebSocket message handling to include ping/pong for connection health checks.
- Create type safe websockets with shared types in libs/ws/index

lbxa commented on apps/backend/src/ws/socketHandler.ts:
```diff
@@ -23,24 +32,84 @@ export class SocketHandler {
+
+          // Validate that parsed data matches WebSocketLocationUpdate structure
+          if (this.isValidLocationUpdate(parsed)) {
+            for (const socketId of this.locationUpdatesRoom) {
```
I feel sorry for the server running this once we have a few thousands requests per minute lol

jackrdye replied:
🕺

lbxa commented on apps/backend/src/ws/socketHandler.ts:
```diff
@@ -189,12 +250,71 @@ export class SocketHandler {
+        staleConnections.push(socketId);
+
+        if (ws.readyState === WebSocket.OPEN) {
+          ws.close(1001, "Connection timeout");
```
What's the magic number `1001` here for?

lbxa commented on apps/backend/src/ws/socketHandler.ts:
```diff
@@ -208,6 +328,18 @@ export class SocketHandler {
+    // Unsubscribe and disconnect Redis subscriber
+    try {
+      if (this.redisSubscriber) {
+        try {
```
nitpicking: worth re-writing as 

```ts
try {
  await this.redisSubscriber.unsubscribe(this.redisChannel);
} catch (_) {
  // nothing
} finally {
  await this.redisSubscriber.disconnect()
}
```

lbxa reviewed (changes requested):
Lgtm - a few small things

jackrdye committed:
refactor: Add 'hours' query string to getAllLocations request. Hours determines the number of hours back you want all location data for (from the current time).

jackrdye commented on apps/backend/src/trpc/routers/location.ts:
```diff
@@ -1,29 +1,23 @@
+  getAll: publicProcedure.output(LocationsResponseSchema).query(async ({ ctx }) => {
+    try {
+      const result = await ctx.locationService.getLocations();
+      return {
```
Done - updated commit

lbxa committed:
feat: integrate device ID handling for location filtering in App component

- Added device ID service to retrieve the current device ID for filtering timeline locations.
- Updated the feature mapping logic to exclude locations associated with the current device ID.
- Enhanced dependency array in useEffect to include device ID for accurate updates.
- Improved code readability by formatting the setPaintProperty method in TimelineController.

jackrdye committed:
refactor: improve error handling in Redis subscriber disconnection process

- Simplified the unsubscribe and disconnect logic for the Redis subscriber.
- Added error handling to log issues during the disconnection process while ensuring the subscriber is properly unsubscribed and disconnected.

lbxa committed:
small ittle things

lbxa committed:
toasting

lbxa committed:
remove comment

lbxa committed:
quick commit

lbxa committed:
removing dockerfile from railway
